### DISCUSSION TITLE
jts-core version upgrade?

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/vtm#935] [Ideas]
Hi @devemux86,

dependabot recently informed us that version 0.19.0 of jts-core has become available. We are currently using 0.16.1 in conjunction with vtm / c:geo (can't remember anymore why we selected that specific version, though), and the integration guide still states 0.15.1, so I'm a bit confused.

I did a quick test by changing build.gradle, but then the map crashes with an exception. A quick search for this brought up the reason: There has been a slight interface change with 0.17 (?), as described in [this issue by a different project using jts](https://github.com/wdtinc/mapbox-vector-tile-java/issues/48), which leads to this crash. Adapting to this would require a slight code change in VTM.

Regarding your comment in [this old vtm issue](https://github.com/mapsforge/vtm/issues/738) I would expect this to be the compileSDK setting, which currently needs to be at API level 30 anyway, IIRC, according to Play Store rules. This setting is independent from min supported API version. c:geo still runs on API level 21+ currently. I did a quick test with setting jts-core version number to 0.19.0 and compiling this for an emulated API 21 device - compiles fine and starts without problems (but crashes on opening the vtm-based map due to the interface change described above).

Do you have any plans to adapt vtm to most recent versions of jts (maybe under what conditions) or would you accept a PR for it?

**Comment by devemux86** [comment]
The reason is linked in the [integration guide](https://github.com/mapsforge/vtm/blob/master/docs/Integration.md#jts-geometries):
https://github.com/locationtech/jts/issues/145

[Mapsforge](https://github.com/mapsforge/mapsforge/blob/master/build.gradle#L33-L34) and [VTM](https://github.com/mapsforge/vtm/blob/master/build.gradle#L35-L36) libraries use:
```groovy
sourceCompatibility = JavaVersion.VERSION_1_7
targetCompatibility = JavaVersion.VERSION_1_7
```

New JTS (and [OkHttp](https://github.com/square/okhttp/issues/4481)) seem to require Java 8.

**Reply by moving-bits** [reply]
Thanks for the quick reply. Is there any reason to still stay with Java 7 nowadays? As long as you only use Java 8 syntax and avoid the new API, it should still work, as the compiler transforms the syntactic sugar. c:geo has switched to Java 8 a few years ago (and 4.x series of ok.http, BTW), and we haven't noticed any problems related to it yet. (We are supporting Android API 21+ currently, the switch may have even been in the API 20+ time already, not sure.)

Is there any chance to make vtm code compatible with both?

**Comment by devemux86** [comment]
> Is there any reason to still stay with Java 7 nowadays?

There were reports with Java 8 problems on some Androids and devices.
Java 7 exists for maximum compatibility, without valid reason for upgrade.

I may update it in the future, but after _testing_.
(library development is not based on bots)

Updates of JTS / OkHttp may be better or not.
Map libraries make little use of their large API.

> Is there any chance to make vtm code compatible with both?

Mapsforge and VTM are open source libraries.
Anyone can build them with different parameters. :slightly_smiling_face:

**Reply by moving-bits** [reply]
Thanks for the explanation. Fully understandable from the library point of view.

From the apps perspective I mostly try to be more or less up-to-date with (most) dependencies due to bug fixes of all kind, or for compatibility reasons, and sometimes for functional improvements. dependabot helps me with detecting available updates, nothing more, and it's up to the maintainers to decide which update to include or not. And that's why I opened this thread, to ask you about this dependency, as jts is currently used by vtm only in our context. 
I'll probably adapt dependabot config to be happy with 0.16.1 for now.

Thanks for your constant support!
